mmselfsup/models/heads/mae_head.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcls.models import LabelSmoothLoss
+from mmcv.cnn.utils.weight_init import trunc_normal_
+from mmcv.runner import BaseModule
+from torch import nn
+
+from ..builder import HEADS
+
+
+@HEADS.register_module()
+class MAEPretrainHead(BaseModule):
+    """Pre-training head for MAE.
+
+    Args:
+        norm_pix_loss (bool): Whether or not normalize target.
+            Defaults to False.
+        patch_size (int): Patch size. Defaults to 16.
+    """
+
+    def __init__(self, norm_pix=False, patch_size=16):
+        super(MAEPretrainHead, self).__init__()
+        self.norm_pix = norm_pix
+        self.patch_size = patch_size
+
+    def patchify(self, imgs):
+
+        p = self.patch_size
+        assert imgs.shape[2] == imgs.shape[3] and imgs.shape[2] % p == 0
+
+        h = w = imgs.shape[2] // p
+        x = imgs.reshape(shape=(imgs.shape[0], 3, h, p, w, p))
+        x = torch.einsum('nchpwq->nhwpqc', x)
+        x = x.reshape(shape=(imgs.shape[0], h * w, p**2 * 3))
+        return x
+
+    def forward(self, x, pred, mask):
+        losses = dict()
+        target = self.patchify(x)
+        if self.norm_pix:
+            mean = target.mean(dim=-1, keepdim=True)
+            var = target.var(dim=-1, keepdim=True)
+            target = (target - mean) / (var + 1.e-6)**.5
+
+        loss = (pred - target)**2
+        loss = loss.mean(dim=-1)
+
+        loss = (loss * mask).sum() / mask.sum()
```

Comment:
We can use mask to select the pred and target first then calculate the loss to reduce the memory usage?